source/common/stats/thread_local_store.cc

```diff
@@ -61,6 +61,27 @@ std::list<GaugeSharedPtr> ThreadLocalStoreImpl::gauges() const {
   return ret;
 }
 
+std::list<HistogramSharedPtr> ThreadLocalStoreImpl::histograms() const {
+  // Handle de-dup due to overlapping scopes.
+  std::list<HistogramSharedPtr> ret;
+  std::unordered_set<std::string> names;
+  std::unique_lock<std::mutex> lock(lock_);
+  // TODO(ramaraochavali): incorporate the scopes into the histogram names.
+  for (ScopeImpl* scope : scopes_) {
+    for (auto histogram : scope->central_cache_.histograms_) {
+      const std::string& hist_name = histogram.first;
+      const ParentHistogramSharedPtr& parent_hist = histogram.second;
+      if (names.insert(hist_name).second) {
+        ret.push_back(parent_hist);
+      } else {
+        ENVOY_LOG(warn, "duplicate histogram name '{}'", hist_name);
```

Comment:
Thanks for adding the warning; you might want to also throw in the scope name.


I'm still interested in why that occurs during hot restart.  It doesn't seem obvious to me that there should be duplicate scopes or duplicate histograms within each scope during hot restart. 

The fact that this occurs only in the hot restart tests suggests to me that the TODO above does not cover the issue.  It suggests to me that this is a data model issue on hot restart.

Reply:
I added scope to the log line and also added the log line for counters and gauges  to see if they have duplicates during hot restart. It seems that they also have duplicates. So it is not unique to histograms. You can grab any of the test output in CI and see the duplicate counters, gauges and histogram list.